internal/identity/idm_test.go

```diff
@@ -0,0 +1,330 @@
+// Copyright 2015 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package identity_test
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"mime"
+	"net/http"
+	"net/http/httptest"
+	"strconv"
+	"strings"
+
+	"github.com/juju/utils"
+	gc "gopkg.in/check.v1"
+	"gopkg.in/macaroon-bakery.v1/bakery"
+	"gopkg.in/macaroon-bakery.v1/bakery/checkers"
+	"gopkg.in/macaroon-bakery.v1/httpbakery"
+
+	"gopkg.in/juju/charmstore.v5-unstable/internal/identity"
+)
+
+type discharge struct {
+	id string
+	c  chan error
+}
+
+type idM struct {
+	*httptest.Server
+	*http.ServeMux
+	svc        *bakery.Service
+	discharges map[string]discharge
+	key        *bakery.KeyPair
+}
+
+func NewIdM(c *gc.C) *idM {
+	i := &idM{
+		ServeMux:   http.NewServeMux(),
+		discharges: make(map[string]discharge),
+	}
+	i.Server = httptest.NewServer(i)
+	var err error
+	i.key, err = bakery.GenerateKey()
+	c.Assert(err, gc.IsNil)
+	i.svc, err = bakery.NewService(bakery.NewServiceParams{
+		Key: i.key,
+		Locator: bakery.PublicKeyLocatorMap{
+			i.URL: &i.key.Public,
+		},
+	})
+	c.Assert(err, gc.IsNil)
+	httpbakery.AddDischargeHandler(i.ServeMux, "/", i.svc, i.checker)
+	i.Handle("/", http.HandlerFunc(i.notFound))
+	i.Handle("/test", http.HandlerFunc(i.serveTest))
+	i.Handle("/protected", http.HandlerFunc(i.serveProtected))
+	i.Handle("/login", http.HandlerFunc(i.serveLogin))
+	i.Handle("/wait", http.HandlerFunc(i.serveWait))
+	i.Handle("/agent", http.HandlerFunc(i.serveAgent))
+	i.Handle("/v1/u/user1/groups", i.serveGroups("g1", "g2"))
+	i.Handle("/v1/u/user2/groups", i.serveGroups())
+	return i
+}
+
+func (i *idM) notFound(w http.ResponseWriter, req *http.Request) {
+	i.error(w, http.StatusNotFound, "not found", "%s not found", req.URL.Path)
+}
+
+func (i *idM) serveTest(w http.ResponseWriter, req *http.Request) {
```

Comment:
As discussed, it would be nice to have some doc comments on the endpoints here to remind us what functionality they implement and how that may or may not related to the actual idm endpoints.